x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/BaseResponseHandler.java

```diff
@@ -54,11 +59,38 @@ public String getRequestType() {
         return requestType;
     }
 
+    @Override
+    public void validateResponse(ThrottlerManager throttlerManager, Logger logger, Request request, HttpResult result) {
+        checkForFailureStatusCode(request, result);
+        checkForEmptyBody(throttlerManager, logger, request, result);
+
+        // When the response is streamed the status code could be 200 but the error object will be set
+        // so we need to check for that specifically
+        checkForErrorObject(request, result);
+    }
+
+    protected abstract void checkForFailureStatusCode(Request request, HttpResult result);
+
+    private void checkForErrorObject(Request request, HttpResult result) {
+        var errorEntity = errorParseFunction.apply(result);
+
+        if (errorEntity.errorStructureFound()) {
+            // we don't really know what happened because the status code was 200 so we'll just retry
+            throw new RetryException(true, buildError(SERVER_ERROR_OBJECT, request, result, errorEntity));
```

Comment:
Should we just retry if we get a failure mid-stream?

Reply:
Looking at the OpenAI java and python libraries, it seems like they don't retry on error messages. 
https://github.com/openai/openai-java/blob/9a6fa4aeea6738190dc1eb28cd52d82fbbf0b18b/openai-java-core/src/main/kotlin/com/openai/core/handlers/SseHandler.kt#L58
https://github.com/openai/openai-python/blob/995cce048f9427bba4f7ac1e5fc60abbf1f8f0b7/src/openai/_streaming.py#L70

The SSE spec does seem to include some allowance for retries, but I'm not sure that it allows for retry on an error returned from the EventSource. It seems that retries mid-stream might only be for disconnections. I'm not totally sure though. https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation

Reply:
Great points. I'll switch it to skip retrying. I think that's also fine for the non-streaming case too. If we are in the non-streaming code path I'd expect a failure status code first which would cause a throw so we shouldn't even be getting to this code block in that scenario.